board/sunnypilot/safety_mads.h

```diff
@@ -0,0 +1,173 @@
+      button_state->last == MADS_BUTTON_PRESSED
+    );
+
+    button_state->last = *button_state->current;
+  }
+}
+
+// Main cruise
+static void m_update_binary_state(BinaryStateTracking *state) {
+  state->transition = m_get_edge_transition(*state->current, state->previous);
+  state->previous = *state->current;
+}
+
+static void m_mads_try_allow_controls_lat(void) {
+  if (m_mads_state.controls_requested_lat && !m_mads_state.controls_allowed_lat && m_can_allow_controls_lat()) {
+    m_mads_state.controls_allowed_lat = true;
+    m_mads_state.previous_disengage = m_mads_state.current_disengage;
+    m_mads_state.current_disengage.reason = MADS_DISENGAGE_REASON_NONE;
+  }
+}
+
+// Use buttons or main cruise state transition properties to request lateral control
+static void m_mads_update_state(void) {
+  // Main cruise
+  if (m_mads_state.acc_main.transition == MADS_EDGE_RISING) {
+    m_mads_state.controls_requested_lat = true;
+  } else if (m_mads_state.acc_main.transition == MADS_EDGE_FALLING) {
+    m_mads_state.controls_requested_lat = false;
+    mads_exit_controls(MADS_DISENGAGE_REASON_ACC_MAIN_OFF);
+  } else {
+  }
+
+  // MADS button
+  if (m_mads_state.mads_button.transition == MADS_EDGE_RISING) {
+    m_mads_state.controls_requested_lat = !m_mads_state.controls_allowed_lat;
+
+    if (!m_mads_state.controls_requested_lat) {
+      mads_exit_controls(MADS_DISENGAGE_REASON_BUTTON);
+    }
+  }
+
+  if (m_mads_state.op_controls_allowed.transition == MADS_EDGE_RISING) {
+    m_mads_state.controls_requested_lat = true;
+  }
+}
+
+// ===============================
+// Function Implementations
+// ===============================
+
+inline const MADSState *get_mads_state(void) {
+  return &m_mads_state;
+}
+
+inline void mads_set_system_state(bool enabled, bool disengage_lateral_on_brake) {
+  m_mads_state_init();
+  m_mads_state.system_enabled = enabled;
+  m_mads_state.disengage_lateral_on_brake = disengage_lateral_on_brake;
+}
+
+inline void mads_exit_controls(DisengageReason reason) {
+  if (m_mads_state.controls_allowed_lat) {
+    m_mads_state.previous_disengage = m_mads_state.current_disengage;
+    m_mads_state.current_disengage.reason = reason;
+    m_mads_state.controls_allowed_lat = false;
+  }
+}
+
+inline bool mads_is_lateral_control_allowed_by_mads(void) {
+  return m_mads_state.system_enabled && m_mads_state.controls_allowed_lat;
+}
+
+inline void mads_state_update(const bool *op_vehicle_moving, const bool *op_acc_main, const bool *op_allowed, bool is_braking) {
+  m_mads_state.is_vehicle_moving_ptr = op_vehicle_moving;
+  m_mads_state.acc_main.current = op_acc_main;
+  m_mads_state.op_controls_allowed.current = op_allowed;
+  m_mads_state.mads_button.current = &mads_button_press;
+
+  m_update_binary_state(&m_mads_state.acc_main);
+  m_update_binary_state(&m_mads_state.op_controls_allowed);
+  m_update_button_state(&m_mads_state.mads_button);
+
+  m_mads_update_state();
+
+  m_mads_check_braking(is_braking);
```

Comment:
this one feels closer to a `m_update_binary_state`. when I created `m_mads_check_braking` I hadn't made yet the binary state track, but now that we have it maybe we can just switch to it for more consistency? This is not really needed to feel free to ignore it. Once we have this working in general and with all the tests refactoring parts like this should be a walk in the park.

Reply:
We can revisit this in a future PR. Added TODO.
```suggestion
  // TODO-SP: Refactor to utilize `m_update_binary_state`
  m_mads_check_braking(is_braking);
```